torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  is not the same as any other array within the input iterable OR if its shape does not match any other
+  array's shape in the iterable.
+  */
+
+  PyArrayObject* array;
+  if (PyArray_Check(obj)) {
+    array = (PyArrayObject*) obj;
+  } else {
+    return 0;
+  } 
+
+  if (PyArray_NDIM(array) == 0) {
+    return -1;
+  }
+
+  if (ndarray_shape.size() == 0) {
+    /*
+    If ndarray_shape is an empty vector, then ndarray_shape[0] is set to the depth of the array in the iterable and
+    ndarray_shape[1:] is set to the shape of array.
+    */
+    ndarray_shape.emplace_back(depth);
+    int64_t ndim = PyArray_NDIM(array);
+    npy_intp* array_shape = PyArray_DIMS(array);
+    ndarray_shape.insert(ndarray_shape.end(), array_shape, array_shape + ndim);
+  }
+  else if ((ndarray_shape[0] != depth) || !_shape_match_ndarray(ndarray_shape, array, 1)) {
+    /*
+    If the depth of 'array' in the iterable is not the same as any other array's depth OR if the shape of 'array' does not
+    match with any other array's shape in the iterable, we raise a ValueError.
+    */
+    auto true_shape_str = get_buffer_str<npy_intp> (static_cast<void*>(PyArray_DIMS(array)), 0, PyArray_NDIM(array));
+    auto expected_shape_str = get_vector_str(ndarray_shape, 1);
+    std::ostringstream err;
+    err << "expected numpy array of shape " << expected_shape_str <<" at dim " << ndarray_shape[0] << " (got " << true_shape_str << " at dim " << depth << ")";
+    throw ValueError(err.str());  
+  }
+
+  return 1;
+}
+
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t depth, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  This is a helper function for extracting all the numpy arrays embedded within PySequences (recursively) representd by obj. This is done recursively in a 
+  depth first search style. It puts the numpy array pointers in the array_ptr_storage for later use. In case there is a non numpy array type within obj, 
+  this function immediately returns false. If only all the items inside obj are just non-scalar numpy arrays, it returns true. It throws a ValueError in case
+  of a dimension and/or depth mismatch (See _validate_ndarray for more info). This function is used in extract_ndarrays() function below.
+  */
+
+  int validation_flag = _validate_ndarray(obj, shape, ndarray_shape, depth);
+  if (validation_flag != 0) {
+    if (validation_flag == -1) { // If obj is a scalar array, fall back to the slower path.
+      return false;
+    }
+    array_ptr_storage.push_back((PyArrayObject*) obj);
+    return true;
+  } else if (PySequence_Check(obj)) {
+    auto seq = THPObjectPtr(PySequence_Fast(obj, "not a sequence"));
+    if (!seq) {
+      throw python_error();
+    }
+    Py_ssize_t seq_len = PySequence_Fast_GET_SIZE(seq.get());
+    if (seq_len != shape[depth]) {
+      throw ValueError("expected sequence of length %lld at dim %lld (got %lld)",
```

Comment:
Same comment here as above about using the existing macros vs. directly throwing errors